.circleci/config.yml

```diff
@@ -213,6 +213,112 @@ jobs:
+              -e 'cat(sprintf("export DT_TARBALL=%s_%s.tar.gz\n", dt_dsc[,"Package"], dt_dsc[,"Version"]))' \
+              -e 'cat(sprintf("export DT_CHECK_DIR=%s.Rcheck\n", dt_dsc[,"Package"]))' \
+              >> ${BASH_ENV}
+
+      - run:
+          name: ️️📋run CRAN package checks
+          command: |
+            R CMD build dash-core-components
+            R CMD build dash-html-components
+            R CMD build dash-table
+            R CMD build dashR
+            sudo R CMD INSTALL dash-core-components
+            sudo R CMD INSTALL dash-html-components
+            sudo R CMD INSTALL dash-table
+            sudo R CMD INSTALL dashR
+            R CMD check "${DHC_TARBALL}" --as-cran --no-manual
+            R CMD check "${DCC_TARBALL}" --as-cran --no-manual
+            R CMD check "${DT_TARBALL}" --as-cran --no-manual
+            R CMD check "${DASH_TARBALL}" --as-cran --no-manual
+
+      - run:
+          name: 🕵detect failures
+          command: |
+            Rscript -e "message(devtools::check_failures(path = '${DHC_CHECK_DIR}'))"
+            Rscript -e "message(devtools::check_failures(path = '${DCC_CHECK_DIR}'))"
+            Rscript -e "message(devtools::check_failures(path = '${DT_CHECK_DIR}'))"
+            Rscript -e "message(devtools::check_failures(path = '${DASH_CHECK_DIR}'))"                                    
+            # warnings are errors; enable for stricter checks once CRAN submission finished
+            # if grep -q -R "WARNING" "${DHC_CHECK_DIR}/00check.log"; then exit 1; fi
+            # if grep -q -R "WARNING" "${DCC_CHECK_DIR}/00check.log"; then exit 1; fi
+            # if grep -q -R "WARNING" "${DT_CHECK_DIR}/00check.log"; then exit 1; fi
+            # if grep -q -R "WARNING" "${DASH_CHECK_DIR}/00check.log"; then exit 1; fi
+
+      - run:
+          name: 🔎 run unit tests
+          command: |
+              sudo Rscript -e 'res=devtools::test("dashR/tests/", reporter=default_reporter());df=as.data.frame(res);if(sum(df$failed) > 0 || any(df$error)) {q(status=1)}'
+
+      - run:
+          name: ⚙️ Integration tests
+          command: |
+            python -m venv venv
+            . venv/bin/activate
+            cd dash-main/\@plotly/dash-generator-test-component-nested && npm ci && npm run build && sudo R CMD INSTALL . && cd ../../..
+            cd dash-main/\@plotly/dash-generator-test-component-standard && npm ci && npm run build && sudo R CMD INSTALL . && cd ../../..
+            export PATH=$PATH:/home/circleci/.local/bin/
+            pytest --nopercyfinalize --junitxml=test-reports/dashr.xml dashR/tests/integration/dopsa/
+      - store_artifacts:
+          path: test-reports
+      - store_test_results:
+          path: test-reports
+      - store_artifacts:
+          path: /tmp/dash_artifacts
+
+      - run:
+          name: 🦔 percy finalize
+          command: npx percy finalize --all
+          when: always
```

Comment:
might need this to be `when: on_fail`, and have the `percy_finalize` job require this one in addition to `test-37`? That's still not quite the right logic, as we'll finalize the Percy job twice if either `build-dashr` or `test-37` fails - cutting off any further snapshots in the one that didn't fail. But it's probably the best we can do given that CircleCI doesn't have "require these jobs to be done and execute even if they failed"

Reply:
Ah, interesting. I hadn't considered that these tests could *both* modify when the Percy job is finalized. It does seem like `on_fail` is at least preferable to `always`, though.

Reply:
fixed in 68417096